test/subspace-test-service/src/mock.rs

```diff
@@ -184,8 +184,21 @@ impl MockPrimaryNode {
                 .retain(|subscriber| subscriber.unbounded_send(value.clone()).is_ok());
         }
 
-        while (slot_acknowledgement_receiver.next().await).is_some() {
-            // Wait for all the acknowledgements to progress.
+        // Wait for all the acknowledgements to progress.
+        let mut expected_ack = self.new_slot_notification_subscribers.len();
+        let mut received_ack = 0;
+        while expected_ack != received_ack {
+            select! {
+                res = slot_acknowledgement_receiver.next() => match res {
+                    Some(_) => received_ack += 1,
+                    None => break,
+                },
+                // Update `expected_ack` in case subscriber dropped after notification
+                _ = futures_timer::Delay::new(time::Duration::from_millis(500)).fuse() => {
+                    self.new_slot_notification_subscribers.retain(|subscriber| !subscriber.is_closed());
+                    expected_ack = self.new_slot_notification_subscribers.len();
+                }
+            }
         }
```

Comment:
This feels much more complex than it should be. I think this is doing the same:
```suggestion
        // Wait for all the acknowledgements to progress and proactively drop closed subscribers.
        loop {
            select! {
                res = slot_acknowledgement_receiver.next() => if res.is_none() {
                    break;
                },
                // TODO: Workaround for https://github.com/smol-rs/async-channel/issues/23, remove once fix is released
                _ = futures_timer::Delay::new(time::Duration::from_millis(500)).fuse() => {
                    self.new_slot_notification_subscribers.retain(|subscriber| !subscriber.is_closed());
                }
            }
        }
```

Reply:
You're right, thanks!